bee-network/bee-autopeering/src/task.rs

```diff
@@ -0,0 +1,148 @@
+    shutdown_senders: HashMap<String, ShutdownTx>,
+    peerstore: S,
+    active_peers: ActivePeersList,
+    replacements: ReplacementList,
+}
+
+impl<S: PeerStore, const N: usize> TaskManager<S, N> {
+    pub(crate) fn new(peerstore: S, active_peers: ActivePeersList, replacements: ReplacementList) -> Self {
+        Self {
+            shutdown_order: PriorityQueue::with_capacity(N),
+            shutdown_senders: HashMap::with_capacity(N),
+            shutdown_handles: HashMap::with_capacity(N),
+            peerstore,
+            active_peers,
+            replacements,
+        }
+    }
+
+    /// Runs a `Runnable`, which is a type that features an event loop.
+    pub(crate) fn run<R>(&mut self, runnable: R)
+    where
+        R: Runnable<ShutdownSignal = ShutdownRx> + 'static,
+    {
+        let (shutdown_tx, shutdown_rx) = oneshot::channel::<()>();
+        self.shutdown_senders.insert(R::NAME.into(), shutdown_tx);
+
+        let handle = tokio::spawn(runnable.run(shutdown_rx));
+        log::trace!("`{}` running.", R::NAME);
+
+        assert!(!self.shutdown_handles.contains_key(R::NAME));
+        self.shutdown_handles.insert(R::NAME.into(), handle);
+
+        self.shutdown_order.push(R::NAME.into(), R::SHUTDOWN_PRIORITY);
+    }
+
+    /// Repeats a command in certain intervals provided a context `T`. Will be shut down gracefully with the rest of
+    /// spawned task by specifying a `name` and a `shutdown_priority`.
+    pub(crate) fn repeat<T, D>(&mut self, cmd: Repeat<T>, mut delay: D, ctx: T, name: &str, shutdown_priority: u8)
+    where
+        T: Send + Sync + 'static,
+        D: Iterator<Item = Duration> + Send + 'static,
+    {
+        let (shutdown_tx, mut shutdown_rx) = oneshot::channel::<()>();
+        self.shutdown_senders.insert(name.into(), shutdown_tx);
+
+        let handle = tokio::spawn(async move {
+            for duration in &mut delay {
+                tokio::select! {
+                    _ = &mut shutdown_rx => break,
+                    _ = time::sleep(duration) => cmd(&ctx),
+                }
+            }
+        });
+        log::trace!("`{}` repeating.", name);
+
+        assert!(!self.shutdown_handles.contains_key(name));
+        self.shutdown_handles.insert(name.into(), handle);
+
+        self.shutdown_order.push(name.into(), shutdown_priority);
+    }
+
+    /// Executes the system shutdown.
+    pub(crate) async fn shutdown(self) {
+        let TaskManager {
+            mut shutdown_order,
+            shutdown_handles: mut runnable_handles,
+            mut shutdown_senders,
+            peerstore,
+            active_peers,
+            replacements,
+        } = self;
+
+        // Send the shutdown signal to all receivers.
+        // TODO: clone necessary?
+        let mut shutdown_order_clone = shutdown_order.clone();
+        while let Some((task_name, _)) = shutdown_order_clone.pop() {
+            let shutdown_tx = shutdown_senders.remove(&task_name).unwrap();
```

Comment:
Unhandled unwrap()?

Reply:
Unwrapping can never fail here, but I forgot adding the comment that explains why. Fixed.